solr/solrj/src/java/org/apache/solr/common/util/SimpleOrderedMap.java

```diff
@@ -68,24 +75,123 @@ public SimpleOrderedMap(Map.Entry<String, T>[] nameValuePairs) {
   public SimpleOrderedMap<T> clone() {
     ArrayList<Object> newList = new ArrayList<>(nvPairs.size());
     newList.addAll(nvPairs);
-    return new SimpleOrderedMap<>(newList);
+    return new SimpleOrderedMap<T>(newList);
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return nvPairs.isEmpty();
+  }
+
+  @Override
+  public boolean containsKey(final Object key) {
+    return this.indexOf((String) key) >= 0;
+  }
+
+  @Override
+  public boolean containsValue(final Object value) {
+    return values().contains(value);
   }
 
   /**
-   * Returns a shared, empty, and immutable instance of SimpleOrderedMap.
+   * {@inheritDoc}
+   *
+   * <p>Has linear lookup time O(N)
    *
-   * @return Empty SimpleOrderedMap (immutable)
+   * @see NamedList#get(String)
    */
-  public static SimpleOrderedMap<Object> of() {
-    return EMPTY;
+  @Override
+  public T get(final Object key) {
+    return super.get((String) key);
+  }
+
+  @Override
+  public T put(String key, T value) {
+    int idx = indexOf(key);
+    if (idx == -1) {
+      add(key, value);
+      return null;
+    } else {
+      T t = get(key);
+      setVal(idx, value);
+      return t;
+    }
   }
 
   /**
-   * Returns an immutable instance of SimpleOrderedMap with a single key-value pair.
+   * @see NamedList#remove(String)
+   */
+  @Override
+  public T remove(final Object key) {
+    return super.remove((String) key);
+  }
+
+  @Override
+  public void putAll(final Map<? extends String, ? extends T> m) {
+    if (isEmpty()) {
+      m.forEach(this::add);
+    } else {
+      m.forEach(this::put);
+    }
+  }
+
+  @Override
+  public Map<String, T> asShallowMap() {
```

Comment:
nitpick but place this up above implementing Map interfaces rather than arbitrarily somewhere in the middle.